src/k8s/pkg/k8sd/api/certs_refresh.go

```diff
@@ -0,0 +1,277 @@
+	snap := e.provider.Snap()
+
+	isWorker, err := snaputil.IsWorker(snap)
+	if err != nil {
+		return response.InternalError(fmt.Errorf("failed to check if node is a worker: %w", err))
+	}
+
+	if isWorker {
+		return refreshCertsPlanWorker(s, snap)
+
+	} else {
+		// TODO: Control Plane refresh
+		return response.InternalError(fmt.Errorf("not implemented yet"))
+	}
+
+}
+
+// refreshCertsPlanWorker generates the CSRs for the worker node and returns the seed and the names of the CSRs.
+func refreshCertsPlanWorker(s *state.State, snap snap.Snap) response.Response {
+	client, err := snap.KubernetesNodeClient("")
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	var seed int32
+
+	err = binary.Read(rand.Reader, binary.BigEndian, &seed)
+	if err != nil {
+		return response.InternalError(err)
+	}
+	seed = seed & 0x7FFFFFFF
+
+	csrKubeletServing, pKeyKubeletServing, err := pkiutil.GenerateCSR(
+		pkix.Name{
+			CommonName:   fmt.Sprintf("system:node:%s", snap.Hostname()),
+			Organization: []string{"system:nodes"},
+		},
+		2048,
+		nil,
+		[]string{snap.Hostname()},
+		[]net.IP{net.ParseIP(s.Address().Hostname())},
+	)
+
+	_, err = client.CreateCertificateSigningRequest(
+		s.Context,
+		fmt.Sprintf("k8sd-%d-worker-kubelet-serving", seed),
+		[]byte(csrKubeletServing),
+		[]v1.KeyUsage{v1.UsageDigitalSignature, v1.UsageKeyEncipherment, v1.UsageServerAuth},
+		[]string{"system:nodes"},
+		"k8sd.io/kubelet-serving",
+	)
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	csrKubeletClient, pKeyKubeletClient, err := pkiutil.GenerateCSR(
+		pkix.Name{
+			CommonName:   fmt.Sprintf("system:node:%s", snap.Hostname()),
+			Organization: []string{"system:nodes"},
+		},
+		2048,
+		nil,
+		nil,
+		nil,
+	)
+
+	_, err = client.CreateCertificateSigningRequest(
+		s.Context,
+		fmt.Sprintf("k8sd-%d-worker-kubelet-client", seed),
+		[]byte(csrKubeletClient),
+		[]v1.KeyUsage{v1.UsageDigitalSignature, v1.UsageKeyEncipherment, v1.UsageClientAuth},
+		nil,
+		"k8sd.io/kubelet-client",
+	)
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	csrKubeProxy, pKeyKubeProxy, err := pkiutil.GenerateCSR(
+		pkix.Name{
+			CommonName: "system:kube-proxy",
+		},
+		2048,
+		nil,
+		nil,
+		nil,
+	)
+
+	_, err = client.CreateCertificateSigningRequest(
+		s.Context,
+		fmt.Sprintf("k8sd-%d-worker-kube-proxy-client", seed),
+		[]byte(csrKubeProxy),
+		[]v1.KeyUsage{v1.UsageDigitalSignature, v1.UsageKeyEncipherment, v1.UsageClientAuth},
+		nil,
+		"k8sd.io/kube-proxy-client",
+	)
+
+	result := apiv1.RefreshCertificatesPlanResponse{
+		Seed: int(seed),
+		CertificatesSigningRequests: []string{
+			fmt.Sprintf("k8sd-%d-worker-kubelet-serving", seed),
+			fmt.Sprintf("k8sd-%d-worker-kubelet-client", seed),
+			fmt.Sprintf("k8sd-%d-worker-kube-proxy-client", seed),
+		},
+	}
+
+	err = os.WriteFile(filepath.Join(snap.KubernetesPKIDir(), "kubelet.key.new"), []byte(pKeyKubeletServing), 0600)
```

Comment:
Why are we writing those keys here? I thought the `init/plan` step should be without any side-effects etc.

Reply:
To generate a new CSR, it is necessary to provide a private key. In this case, I chose an approach that creates new keys for the certificates and uses them, instead of reading the existing keys from the kubeconfigs and key files in the respective directory and refreshing all the pki in the node. Since the refresh process is divided into two subprocesses (plan and run), we need to store those keys in a temporary file (hence the .new suffix). This is to ensure that if something goes wrong during the process, we have a backup to keep things running smoothly.

Reply:
Ah that was the context that I missed. Basically, they serve as a backup in case something breaks.

Could you not just do it the other way around in the `run` step:
1. Copy existing certs to `.old`
2. Create new certificates and overwrite the existing PKIs (no need to store in temporary files) 
3. restart services 

That way, if something goes south you can *restore* the old state, something you cannot do if you only store the new certs. If something goes wrong with those certs you are in troubles.

Reply:
Yes, we need to provide a way to back up the existing certificates and keys. However, it is still necessary to create temporary files to save the new signer keys. The process looks like this: create new private keys, sign the CSRs, and save the new private keys in a temporary location. Then, the `run` kicks in, and we wait until the certificates have been signed. Once we have all the certificates, we can replace everything. Changing the keys at this point could potentially block the cluster, as the private keys would not match the old certificates.

Reply:
Three things:

* Do we really want the `plan` step to be mandatory before `run`? In the proposal, the `plan` step is not mandatory and purely informational.
* What happens right now if we forget the `plan` step? I guess the `run` step would break with some cryptic `kubelet.key.new not found` error? IFF we really want to require the user to run `plan` before `run` (as said, I don't thing we should), then at least we need a proper error message to communicate this. 
* The third step in my comment above was wrong. Ofc we need to approve the certificates first, apply them and then we restart the services.

Reply:
1. The `plan` step is [mandatory](https://github.com/canonical/k8s-snap/blob/main/docs/proposals/002-refresh-certificates.md#post-refresh-certsplan) (at least for workers) as this initializes everything required for the `run` endpoint.
> To accomodate this need, while also ensuring that the k8s refresh-certs can be straightforward to implement, two API calls are introduced, which clients should call in order.
...
POST /refresh-certs/plan performs any necessary preparations for certificate refreshes on the current node.

For example, the response we receive from the `plan` endpoint has a seed that we pass later on to the `run` endpoint, allowing us to determine if the CSRs have been approved or not.

2. Yeah, it should break. We can improve the error message to communicate to the user that probably the `plan` step is a prerequisite (again for workers).

Reply:
alright, then let's provide a proper error message and move on.